src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,250 @@
+    }
+    const Matrix<BaseFloat> &RIR = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_rir = RIR.NumCols(),  // number of samples in recording
+          num_chan_rir = RIR.NumRows();  // number of channels in recording
+    KALDI_LOG << "samp_freq of rir is " << samp_freq_rir
+               << " num_samp " << num_samp_rir
+               << " num_chan " << num_chan_rir;
+
+    Vector<BaseFloat> src(num_samp_src);
+    src.CopyRowFromMat(SRC, 0);
+
+    Vector<BaseFloat> rir(num_samp_rir);
+    rir.CopyRowFromMat(RIR, 0);
+    // Assume the rir files are 32 bits
+    rir.Scale(1 / static_cast<float>(2147483647));
+
+    Vector<BaseFloat> noise(0);
+    if (!noise_file.empty()) {
+      float peak = 0;
+      int peak_index = 0;
+      peak = rir.Max(&peak_index);
+      KALDI_LOG << "peak is " << peak;
+      KALDI_LOG << "peak_index is " << peak_index;
+
+      int early_start = peak_index - 0.001 * samp_freq_rir;
+      int early_end = peak_index + 0.05 * samp_freq_rir;
+      if (early_start < 0) early_start = 0;
+      if (early_end > rir.Dim()) early_end = rir.Dim();
+
+      Vector<BaseFloat> early_rir(early_end - early_start);
+      SubVector<BaseFloat> subr(rir, early_start, early_end - early_start);
+      early_rir.CopyFromVec(subr);
+
+      Vector<BaseFloat> early_src(num_samp_src);
+      early_src.CopyFromVec(src);
+      KALDI_LOG << "Run early FFTConvolution ";
+      FFTConvolution(early_src, early_rir);
+
+      // compute the energy
+      for (int32 j = 0; j < early_src.Dim(); j++)
+        early_src(j) *= early_src(j);
+      float src_power = early_src.Sum() / early_src.Dim();
+
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      const Matrix<BaseFloat> &NOISE = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = NOISE.NumCols(),
+            num_chan_noise = NOISE.NumRows();
+      KALDI_LOG << "samp_freq of noise is " << samp_freq_noise
+                << " num_samp " << num_samp_noise
+                << " num_chan " << num_chan_noise;
+      noise.Resize(num_samp_noise);
+      noise.CopyRowFromMat(NOISE, 0);
+
+      // compute the noise energy
+      float n_power_sum = 0;
+      for (int32 j = 0; j < noise.Dim(); j++)
+        n_power_sum += noise(j) * noise(j);
+      float n_power = n_power_sum / noise.Dim();
+
+      KALDI_LOG << "src_power is " << src_power;
+      KALDI_LOG << "n_power is " << n_power;
+
+      float scale_factor = sqrt(pow(10, -snr_db / 10) * src_power / n_power);
+      noise.Scale(scale_factor);
+      KALDI_LOG << "scale_factor: " << scale_factor << " for SNR db " << snr_db;
+    }
+
+    KALDI_LOG << "Run FFTConvolution ";
+    FFTConvolution(src, rir);
+
+    if (noise.Dim() > 0) {
+      RepeatedlyAdd(src, noise);
+    }
+
+    float src_max2 = src.Min();    // for checking the negative max
```

Comment:
move this code to a separate function called

MaxAbsolute(const Vector<BaseFloat> & vector, BaseFloat * max_value)